File: ts/src/marker.ts
import type { ConnectionId } from "./types.js";

interface MarkerData {
  connectionId: ConnectionId;
  connectionToken: bigint;
  acceptedPayloadBytes: bigint;
}

const markerData = new WeakMap<SendMarker, MarkerData>();
let nextConnectionToken = 1n;

/** Opaque boundary after bytes accepted by one connection's local send buffer. */
export class SendMarker {
  private constructor() {}
}

export enum MarkerStatus {
  Pending = "pending",
  Acked = "acked",
  ConnectionGone = "connection-gone",
}

export interface WriteWithMarkerResult {
  accepted: number;
  marker: SendMarker;
}

export class SendProgress {
  private readonly connectionToken = nextConnectionToken++;
  private acceptedPayloadBytes = 0n;
  private ackedPayloadBytes = 0n;

  accept(bytes: number): void {
    this.acceptedPayloadBytes += BigInt(bytes);
  }

  acknowledge(bytes: number): void {
    this.ackedPayloadBytes += BigInt(bytes);
  }

  marker(connectionId: ConnectionId): SendMarker {
    const marker = Object.create(SendMarker.prototype) as SendMarker;
    markerData.set(marker, {
      connectionId,
      connectionToken: this.connectionToken,
      acceptedPayloadBytes: this.acceptedPayloadBytes,
    });
    return Object.freeze(marker);
  }

  status(marker: SendMarker): MarkerStatus {
    const data = markerData.get(marker);
    if (data?.connectionToken !== this.connectionToken) return MarkerStatus.ConnectionGone;
    return this.ackedPayloadBytes >= data.acceptedPayloadBytes
      ? MarkerStatus.Acked
      : MarkerStatus.Pending;
  }
}

export const markerConnectionId = (marker: SendMarker): ConnectionId | undefined =>
  markerData.get(marker)?.connectionId;

File: rust/fips-tcp/src/marker.rs
use std::sync::atomic::{AtomicU64, Ordering};

use crate::types::ConnectionId;

static NEXT_CONNECTION_TOKEN: AtomicU64 = AtomicU64::new(1);

/// Opaque boundary after bytes accepted by one connection's local send buffer.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub struct SendMarker {
    connection_id: ConnectionId,
    connection_token: u64,
    accepted_payload_bytes: u128,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum MarkerStatus {
    Pending,
    Acked,
    ConnectionGone,
}

pub(crate) struct SendProgress {
    connection_token: u64,
    accepted_payload_bytes: u128,
    acked_payload_bytes: u128,
}

impl SendProgress {
    pub(crate) fn new() -> Self {
        let connection_token = NEXT_CONNECTION_TOKEN
            .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |value| {
                value.checked_add(1)
            })
            .expect("TCP/FIPS connection token space exhausted");
        Self {
            connection_token,
            accepted_payload_bytes: 0,
            acked_payload_bytes: 0,
        }
    }

    pub(crate) fn accept(&mut self, bytes: usize) {
        self.accepted_payload_bytes = self
            .accepted_payload_bytes
            .checked_add(bytes as u128)
            .expect("TCP/FIPS accepted payload counter exhausted");
    }

    pub(crate) fn acknowledge(&mut self, bytes: usize) {
        self.acked_payload_bytes = self
            .acked_payload_bytes
            .checked_add(bytes as u128)
            .expect("TCP/FIPS acknowledged payload counter exhausted");
        debug_assert!(self.acked_payload_bytes <= self.accepted_payload_bytes);
    }

    pub(crate) fn marker(&self, connection_id: ConnectionId) -> SendMarker {
        SendMarker {
            connection_id,
            connection_token: self.connection_token,
            accepted_payload_bytes: self.accepted_payload_bytes,
        }
    }

    pub(crate) fn status(&self, marker: &SendMarker) -> MarkerStatus {
        if marker.connection_token != self.connection_token {
            MarkerStatus::ConnectionGone
        } else if self.acked_payload_bytes >= marker.accepted_payload_bytes {
            MarkerStatus::Acked
        } else {
            MarkerStatus::Pending
        }
    }
}

impl SendMarker {
    pub(crate) fn connection_id(self) -> ConnectionId {
        self.connection_id
    }
}
